exchange/auction_test.go

```diff
@@ -94,92 +112,151 @@ func loadCacheSpec(filename string) (*cacheSpec, error) {
-	found := 0
+	_ = testAuction.doCache(ctx, cache, targData, &specData.BidRequest, 60, &specData.DefaultTTLs, bidCategory)
 
-	for _, cExpected := range specData.ExpectedCacheables {
-		for _, cFound := range cache.items {
-			// make sure Data section matches exactly
-			if !vast {
-				// not testing VAST XML for custom cache key test
-				eq := jsonpatch.Equal(cExpected.Data, cFound.Data)
-				if !eq {
-					continue
-				}
-			}
+	if len(specData.ExpectedCacheables) > len(cache.items) {
+		//t.Errorf("%s:  Less elements were cached than expected \n", fileDisplayName)
+		t.Errorf("%s:  [CACHE_ERROR] Less elements were cached than expected \n", fileDisplayName)
+	} else if len(specData.ExpectedCacheables) < len(cache.items) {
+		t.Errorf("%s:  [CACHE_ERROR] More elements were cached than expected \n", fileDisplayName)
+	} else { // len(specData.ExpectedCacheables) == len(cache.items)
+		var ht map[string]*cacheComparator = make(map[string]*cacheComparator)
+		var formattedData string
+		var compareString string
 
-			// make sure Key value is as expected
-			keymatch := false
-			if len(cExpected.Key) > 0 {
-				// can only verify prefix; remainder is random
-				if strings.HasPrefix(cFound.Key, cExpected.Key) {
-					keymatch = true
-				}
-			} else {
-				if len(cFound.Key) == 0 {
-					// Key is expected to be empty
-					keymatch = true
-				}
+		for _, cExpected := range specData.ExpectedCacheables {
+			formattedData = strings.Replace(string(cExpected.Data), "\\", "", -1)
+			formattedData = strings.Replace(formattedData, " ", "", -1)
+			compareString = fmt.Sprintf("%s_%d", formattedData, cExpected.TTLSeconds)
+			// init cacheComparator element before hashing it. If needed
+			if _, ok := ht[compareString]; !ok {
+				ht[compareString] = &cacheComparator{0, make([]string, 0), make([]string, 0)}
 			}
-			if !keymatch {
-				continue
+			ht[compareString].freq += 1
+			if targData.includeCacheVast {
+				ht[compareString].expectedKeys = append(ht[compareString].expectedKeys, cExpected.Key)
 			}
-
-			// make sure TTLSeconds section matches exactly
-			if cExpected.TTLSeconds == cFound.TTLSeconds {
-				found++
+		}
+		for _, cFound := range cache.items {
+			formattedData = strings.Replace(string(cFound.Data), "\\", "", -1)
+			formattedData = strings.Replace(formattedData, " ", "", -1)
+			compareString = fmt.Sprintf("\"%s\"_%d", formattedData, cFound.TTLSeconds)
+			// init cacheComparator element before hashing it. If needed
+			if _, ok := ht[compareString]; !ok {
+				ht[compareString] = &cacheComparator{0, make([]string, 0), make([]string, 0)}
+			}
+			ht[compareString].freq -= 1
+			if targData.includeCacheVast {
+				ht[compareString].actualKeys = append(ht[compareString].actualKeys, cFound.Key)
+			}
+		}
+		for k, cachedElements := range ht {
+			if cachedElements.freq > 0 {
+				t.Errorf("%s:  [CACHE_ERROR] Cache inconsistensy. Element %s was not expected to get cached\n", fileDisplayName, k)
```

Comment:
I just noticed that this error message is confusing. It should say that the element was cached fewer times than expected.  Also, here and two lines below, "inconsistensy" should be "inconsistency".

Reply:
Jeez, the spelling mistake is embarrassing. I just corrected it. I believe that both error messages in lines 230 and 232 are correct aren't they? What changes do you think could make them more descriptive?
```228         for k, cachedElements := range ht {
229             if cachedElements.freq > 0 {
230                 t.Errorf("%s:  [CACHE_ERROR] Cache inconsistency. Element %s was not expected to get cached\n", fileDisplayName, k)
231             } else if cachedElements.freq < 0 {
232                 t.Errorf("%s:  [CACHE_ERROR] Cache inconsistency. We cached some more elements (%s) than expected.\n", fileDisplayName, k)
233             } else if targData.includeCacheVast {
234                 for i := 0; i < len(cachedElements.expectedKeys); i++ {
```

Reply:
Line 230 in your snippet above should say that the element was cached fewer times than expected.

Reply:
What do you think about:
```
t.Errorf("%s:  [CACHE_ERROR] Cache inconsistency. Element '%s' was not cached as was expected. \n", fileDisplayName, k)
``` 
Consider this scenario:
```
Expected      Actual
bidA           bidA
bidB           bidB
bidC
```
This will turn into a hash talbe like this:
```
ht       freq    expectedKeys
bidA360    0          CatA
bidB220    0          CatB
bidC360    1          CatC    //Element 'bidC360' was not cached as we expected
```

In this other scenario:
```
Expected      Actual
bidA           bidA
bidB           bidB
               bidC
```
This will turn into a hash talbe like this:
```
ht       freq    expectedKeys
bidA360    0          CatA
bidB220    0          CatB
bidC360   -1          CatC    //We cached some more elements ('bidC360') than we expected
```